substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -975,9 +982,11 @@ pub mod pallet {
 			#[pallet::compact] max_additional: BalanceOf<T>,
 		) -> DispatchResult {
 			let stash = ensure_signed(origin)?;
-
 			let mut ledger = Self::ledger(StakingAccount::Stash(stash.clone()))?;
 
+			// return early if ledger is in a bad state (ledger's stash != expected).
+			ensure!(ledger.stash == stash, Error::<T>::BadState);
```

Comment:
Ideally move this check inside `fn ledger`?

Reply:
Done!